es5-persistence/src/main/java/com/netflix/conductor/dao/es5/index/ElasticSearchDAOV5.java

```diff
@@ -748,6 +754,47 @@ private void addSortOptionToSearchRequest(SearchRequestBuilder searchRequestBuil
             .collect(Collectors.toCollection(LinkedList::new));
     }
 
+    @Override
+    public void groomWorkflowsAndTasks() {
+        // Groom oldest archived workflows by batch size
+        QueryBuilder wfQuery = QueryBuilders.existsQuery("archived");
+        long workflowsGroomed = groomDocs(indexName, wfQuery, WORKFLOW_DOC_TYPE, "endTime");
+
+        // Groom obsolete tasks that are staying for more than a day
+        int daysToKeepTasks = 1;
+        DateTime dateTime = new DateTime();
+        QueryBuilder taskQuery = QueryBuilders.rangeQuery("updateTime").lt(dateTime.minusDays(daysToKeepTasks));
+
+        long tasksGroomed = groomDocs(indexName, taskQuery, TASK_DOC_TYPE, "updateTime");
+
+        logger.info("Groomed {} workflows and {} tasks", workflowsGroomed, tasksGroomed);
+    }
+
+    public long groomDocs(String indexName, QueryBuilder q, String docType, String sortField) {
+        int batchSize = config.getElasticSearchGroomingBatchSize();
+        SearchRequestBuilder s = elasticSearchClient.prepareSearch(indexName)
+                .setTypes(docType)
+                .setQuery(q)
+                .setSize(batchSize)
+                .addSort(sortField, SortOrder.ASC);
+
+        SearchResponse response = s.execute().actionGet();
+        List<String> docIds = StreamSupport.stream(response.getHits().spliterator(), false)
+                .map(SearchHit::getId)
+                .collect(Collectors.toCollection(LinkedList::new));
+
+        for (String docId : docIds) {
```

Comment:
So the logic will delete batchSize worth of document once a day? And we are deleting one doc at a time. This could be very expensive as we scale up. 
Can we use delete-dy-query and also use async API, so it doesn't block Conductor either.
https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html

Reply:
i thought of that but the current ES API in conductor does not have an easy way to do bulk delete.  DeleteRequest takes index, doctype and id as args but not query. We may have to write a new wrapper.  deleteasync is available though.